modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/CheckboxTag.java

```diff
@@ -0,0 +1,192 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.taglib.clay.servlet.taglib;
+
+import com.liferay.frontend.taglib.clay.internal.servlet.taglib.BaseContainerTag;
+import com.liferay.portal.kernel.language.LanguageUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.taglib.util.TagResourceBundleUtil;
+
+import java.util.Map;
+import java.util.Set;
+
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.JspWriter;
+
+/**
+ * @author Kresimir Coko
+ */
+public class CheckboxTag extends BaseContainerTag {
+
+	@Override
+	public int doStartTag() throws JspException {
+		setAttributeNamespace(_ATTRIBUTE_NAMESPACE);
+
+		return super.doStartTag();
+	}
+
+	public boolean getChecked() {
+		return _checked;
+	}
+
+	public boolean getDisabled() {
+		return _disabled;
+	}
+
+	public boolean getIndeterminate() {
+		return _indeterminate;
+	}
+
+	public boolean getInline() {
+		return _inline;
+	}
+
+	public String getLabel() {
+		return _label;
+	}
+
+	public String getName() {
+		return _name;
+	}
+
+	public void setChecked(boolean checked) {
+		_checked = checked;
+	}
+
+	public void setDisabled(boolean disabled) {
+		_disabled = disabled;
+	}
+
+	public void setIndeterminate(boolean indeterminate) {
+		_indeterminate = indeterminate;
+	}
+
+	public void setInline(boolean inline) {
+		_inline = inline;
+	}
+
+	public void setLabel(String label) {
+		_label = label;
+	}
+
+	public void setName(String name) {
+		_name = name;
+	}
+
+	@Override
+	protected void cleanUp() {
+		super.cleanUp();
+
+		_checked = false;
+		_disabled = false;
+		_indeterminate = false;
+		_inline = false;
+		_label = null;
+		_name = null;
+	}
+
+	@Override
+	protected String getHydratedModuleName() {
+		return "frontend-taglib-clay/Checkbox";
+	}
+
+	@Override
+	protected Map<String, Object> prepareProps(Map<String, Object> props) {
+		props.put("checked", _checked);
+		props.put("disabled", _disabled);
+		props.put("indeterminate", _indeterminate);
+		props.put("inline", _inline);
+
+		if (Validator.isNotNull(_label)) {
+			props.put(
+				"label",
+				LanguageUtil.get(
+					TagResourceBundleUtil.getResourceBundle(pageContext),
+					_label));
+		}
+
+		props.put("name", _name);
+
+		return super.prepareProps(props);
+	}
+
+	@Override
+	protected String processCssClasses(Set<String> cssClasses) {
+		cssClasses.add("custom-control");
+		cssClasses.add("custom-checkbox");
+
+		return super.processCssClasses(cssClasses);
+	}
+
+	@Override
+	protected int processStartTag() throws Exception {
+		super.processStartTag();
+
+		JspWriter jspWriter = pageContext.getOut();
+
+		jspWriter.write("<label><input");
+
+		if (_checked) {
+			jspWriter.write(" checked");
+		}
+
+		jspWriter.write(" class=\"custom-control-input ");
+
+		if (Validator.isNotNull(getCssClass())) {
```

Comment:
I think we have a problem here since we're probably adding `cssClasses` in different ways in different places.

For one, it will be added to the container `div` and now also here with the difference that we now don't go through the whole `processCssClasses` wrappper.

Reply:
Here's the correct place to render those `cssClass` and `id`. If you take a look to the Clay Checkbox component they're rendering `className` and `id` in the `input` and receiving a different param `containerProps` for the main element, so we probably need to overwrite `processCssClasses` or sth to avoid duplication

Reply:
I'll have to recheck that yea, good point.

Reply:
I don't see this being done... aren't we duplicating classes now?

One thing to take into account is that in these cases the semantics we might want to try to maintain are those of `clay:*` tags, not those of `<Clay* />` components. 

If in every other tag, setting `cssClass` implies the main container takes the class and `id` goes to the container but we do it differently here, we're making tags unpredictable.

It's obvious that for `clay:checkbox` our main element should be the `input` and it somehow stands to reason that it should be the one being targeted here... but that has a cost that needs to be evaluated...

Reply:
There are many other things that if we do this would need to target the `input` element as well, such as dynamic attributes... but it's not so obvious because where would you expect `aria-*` or `data-*` to go? The `input` element? The `div` container? somewhere else?

Reply:
@jbalsas I didn't see these comments earlier, GitHub being GitHub I guess. This seems to me like a pretty big hurdle to get over, from my testing I haven't seen any duplications of CSS classes, and all attributes (including classes) are being properly set as they're supposed to when referencing [Checkbox.tsx](https://github.com/liferay/clay/blob/aefb430bd83de2056122ae2af3fb06cae0b2f6ae/packages/clay-form/src/Checkbox.tsx).

The only thing duplicating now is the `id` but, according to @carloslancha, or at least how I understood it, we need to fix that in the base class.

So, what you're suggesting is to alter the wrappers inside the Checkbox tag, so they're consistent with other tags? Something like having the `div` be outside of the actual Checkbox tag, so that the root element of the tag (most likely `input`) is the one that receives the attributes.

That's how I understood it, at least, and I don't really know how to do that, perhaps by making a `BaseCheckboxTag.java` class?